src/lib/logs/gcp_client.py

```diff
@@ -0,0 +1,89 @@
+#   Copyright 2024 Dynatrace LLC
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import asyncio
+import json
+from typing import Any, Dict, List, Callable
+
+from lib.clientsession_provider import init_gcp_client_session
+from lib.context import LoggingContext
+from lib.credentials import create_token
+from lib.logs.log_forwarder_variables import (
+    LOGS_SUBSCRIPTION_ID,
+    LOGS_SUBSCRIPTION_PROJECT,
+    PROCESSING_WORKER_PULL_REQUEST_MAX_MESSAGES,
+)
+
+SUBSCRIPTION_PATH = f"projects/{LOGS_SUBSCRIPTION_PROJECT}/subscriptions/{LOGS_SUBSCRIPTION_ID}"
+
+
+class GCPClient:
+    subscription_path: str
+    pull_url: str
+    acknowledge_url: str
+    body_payload: bytes
+    headers: Dict[str, Any]
+
+    def __init__(
+        self,
+        api_token: str,
+    ):
+        self.subscription_path = (
+            f"projects/{LOGS_SUBSCRIPTION_PROJECT}/subscriptions/{LOGS_SUBSCRIPTION_ID}"
+        )
+        self.pull_url = f"https://pubsub.googleapis.com/v1/{SUBSCRIPTION_PATH}:pull"
+        self.acknowledge_url = f"https://pubsub.googleapis.com/v1/{SUBSCRIPTION_PATH}:acknowledge"
+        self.headers = {"Authorization": f"Bearer {api_token}"}
+        json_body = {"maxMessages": PROCESSING_WORKER_PULL_REQUEST_MAX_MESSAGES}
+        json_data = json.dumps(json_body)
+        self.body_payload = json_data.encode("utf-8")
+
+    async def pull_messages(
+        self, logging_context: LoggingContext, gcp_session, update_gcp_client: Callable[[LoggingContext], None]
+    ) -> Dict[str, List[Any]]:  # type: ignore
+        async with gcp_session.request(
+            method="POST", url=self.pull_url, data=self.body_payload, headers=self.headers
+        ) as response:
+            response_json = await response.json()
+            resp_status = response.status
+
+            if resp_status == 401:
+                await update_gcp_client(logging_context)
```

Comment:
We are sending several pull_messages() tasks, but are we sending them all with the same token? Can they all fail and run update_gcp_client() many times? And we won't pull any messages in that loop. Anyway, just asking, not sure if that can happen.

Reply:
yes, multiple update of the token might happen.

Reply:
If that can happen, it will happen every time we need to update it, right? 
So, each time the token expires, we update it NUMBER_OF_CONCURRENT_MESSAGE_PULL_COROUTINES times? 👀

Reply:
What if in log_integration_service, in the loop from line 55, we do the update once (making sure of this, extra variable or anything) if we see the specific response stating that the token expired?
Plus, we avoid sending update_gcp_client as param to pull_messages().